src/main/java/edu/ucsb/cs156/happiercows/entities/Commons.java

```diff
@@ -1,34 +1,54 @@
 package edu.ucsb.cs156.happiercows.entities;
 
+import java.util.Date;
+import java.util.List;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+
+import javax.persistence.*;
+
 import lombok.Data;
 import lombok.AllArgsConstructor;
 import lombok.NoArgsConstructor;
 import lombok.Builder;
 import lombok.AccessLevel;
 
-import com.fasterxml.jackson.annotation.JsonIgnore;
-
-import java.util.List;
-
-import javax.persistence.*;
-
 @Data
 @AllArgsConstructor
 @NoArgsConstructor(access = AccessLevel.PROTECTED)
 @Builder
 @Entity(name = "commons")
-public class Commons {
+public class Commons
+{
   @Id
   @GeneratedValue(strategy = GenerationType.IDENTITY)
-  private long id;  
+  private long id;
+
   private String name;
+  private double cowPrice;
+  private double milkPrice;
+  private double startingBalance;
+  private Date startingDate;
 
   @ManyToMany(fetch = FetchType.EAGER, cascade = {CascadeType.PERSIST,CascadeType.REMOVE})
-  @JoinTable(name = "user_commons", 
-    joinColumns = @JoinColumn(name = "commons_id", referencedColumnName = "id"), 
+  @JoinTable(name = "user_commons",
+    joinColumns = @JoinColumn(name = "commons_id", referencedColumnName = "id"),
     inverseJoinColumns = @JoinColumn(name = "user_id", referencedColumnName = "id"))
-    @JsonIgnore // https://www.baeldung.com/jackson-bidirectional-relationships-and-infinite-recursion
-    private List<User> users;
-
+  @JsonIgnore // https://www.baeldung.com/jackson-bidirectional-relationships-and-infinite-recursion
+  private List<User> users;
+
+  @Override
+  public int hashCode() {
+    Long boxed = id;
+    return boxed.hashCode();
+  }
+
+  @Override
+  public boolean equals​(Object other) {
+    if (other instanceof Commons) {
+      return this.hashCode() == other.hashCode();
+    }
+
+    return false;
+  }
```

Comment:
Is there any reason why we're overriding `hashCode` and `equals` here? Especially since I don't think we do this in other entities, it might be clearer to do an explicit comparison of ids wherever id-based equality is needed.

Reply:
I made that decision because of this call in the controller tests. I think the argument used to be `any()`, which was bad.
```
verify(commonsRepository, times(1)).save(commons);
```
The `Commons` constructed by the controller is compared (`.equals()`?) to the local `Commons` in the test, which will fail because they're different objects on the heap.

I can add a method to `Commons` to fetch its `id` member if that's clearer.

Reply:
`Commons` is annotated with `@Data`, so there should already be an autogenerated `getId()` method.

As for the test issue, I'd say we arguably want to test more than just "a commons with the same id was saved". After all, what if we save a commons, but we don't save any of the properties that we want changed

Does the `verify(commonsRepository, times(1)).save(commons);` line not work as-is without the `equals`/`hashCode` change? Lombok should autogenerate an appropriate `equals`/`hashCode` for data classes. If we need more granular control over the tests, we can use Mockito's `ArgumentCaptor` or `argThat` matcher (which we don't have any examples of right now, so it would be nice to get them into the codebase!)

Reply:
Yeah, it doesn't work by default.

<img width="1030" alt="diff" src="https://user-images.githubusercontent.com/24307788/156862370-09ecb94a-c6df-4b39-8218-e539a33c06a4.png">

Reply:
Lombok's annotation doesn't seem to work here. Although I do have a typo in this commit; I've corrected it.

```
[ERROR] Failures: 
[ERROR]   CommonsControllerTests.createCommonsTest:98 
Argument(s) are different! Wanted:
edu.ucsb.cs156.happiercows.repositories.CommonsRepository#0 bean.save(
    Commons(id=0, name=Jackson's Commons, cowPrice=500.99, milkPrice=8.99, startingBalance=1020.1, startingDate=Wed Mar 03 16:51:00 PST 2021, users=null)
);
-> at edu.ucsb.cs156.happiercows.controllers.CommonsControllerTests.createCommonsTest(CommonsControllerTests.java:98)
Actual invocations have different arguments:
edu.ucsb.cs156.happiercows.repositories.CommonsRepository#0 bean.save(
    Commons(id=0, name=Jackson's Commons, cowPrice=500.99, milkPrice=8.99, startingBalance=1020.1, startingDate=Wed Mar 03 16:51:00 PST 2021, users=null)
);
-> at edu.ucsb.cs156.happiercows.controllers.CommonsController.createCommons(CommonsController.java:102)
```

Reply:
This is just a guess since I'm not entirely familiar with Mockito internals/semantics, but maybe we need to use `save(eq(commons))` instead of `save(commons)`? Perhaps Mockito defaults to referential equality instead of structural equality when a matcher isn't passed. If that doesn't work, we can fallback to using other Mockito matchers.

Reply:
`eq()` doesn't seem to help.

Reply:
Using `LocalDateTime` fixed this. I have no idea why.